### PR TITLE
[WIP] Set absolute sonames

### PR DESCRIPTION
**Pull request by lheckemann**
###### Motivation for this change
#24844

Most stuff I've tried building with this works. However, nixos won't build with this yet because the extra-tools initrd stuff doesn't build successfully. I'm getting the… interesting error message

```
Inconsistency detected by ld.so: dl-version.c: 205: _dl_check_map_versions: Assertion `needed != NULL' failed!
```

when it tries to test the udevadm in extra-utils. Help would be appreciated!

Also, thanks very much to @grahamc for having made the aarch64 community builder available. It's been invaluable for testing this, what with all the gcc rebuilds!

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/build-support/setup-hooks/set-soname.sh [hunk 1]
```diff
@@ -0,0 +1,8 @@
+[[ -z "$noAbsoluteSoname" ]] && fixupOutputHooks+=(_doAbsoluteSoname)
```
[the remark this answers is not in the document]
Comment (lheckemann) [reply pkgs/build-support/setup-hooks/set-soname.sh]: Done!

**Comment by lheckemann** [comment]
This should also only run on linux, I just realised. Not sure how exactly to do this though — it should probably be in `pkgs/stdenv/linux` somewhere rather than `pkgs/stdenv/generic/default.nix` and guarded by `isLinux`?

**Comment by lheckemann** [comment]
Not much at all. This change doesn't affect dlopen.

**Comment by Mic92** [comment]
I suppose we will uncover bugs, where we implicitly covered dlopen with our rpath? In any case we would also be closer to the macOS port so we have the same behavior on both platforms, which is a good thing.

**Comment by dezgeg** [comment]
I presume this breaks `LD_LIBRARY_PATH`?

IIRC at some point I tried a similar thing which instead of making the `DT_SONAME` directly absolute, it would only `patchelf` all `DT_NEEDED` to absolute paths. I think changing `DT_SONAME` itself makes me slightly more nervous given that AFAICT `DT_SONAME` is sometimes used as some sort of 'cookie' (e.g. if you try to load both `/nix/store/foo/libsomething.so` and `/nix/store/bar/libsomething.so`, if the both have the same `DT_SONAME` only one of them will be loaded).

But, that solution breaks `LD_LIBRARY_PATH` just as well. So nowadays I think the easiest way to solve the `O(n^2)` lookup problem would be to cache the contents of directories in RPATH in the dynamic linker.

**Comment by Mic92** [comment]
When a library was found in the dependency closure and its soname is replaced by an absolute path in the binary, why would we still need LD_LIBRARY_PATH for the same library?

**Comment by lheckemann** [comment]
https://github.com/NixOS/nixpkgs/issues/24844#issuecomment-293664328

We can still use LD_PRELOAD when necessary.

**Comment by lheckemann** [comment]
I had previously rebased this on staging but hadn't actually tried building it after the rebase. I've rebased it on master again because it fails to build a lot more than just the initrd on current staging.

**Comment by edolstra** [comment]
@lheckemann Could you fix that? It's hard to review in the current state.

How does this handle libraries like `libGL.so` that need to be loaded from `LD_LIBRARY_PATH`?

**Comment by vcunat** [comment]
Fixed the GitHub UI.

**Comment by dtzWill** [comment]
This seems to rebase pretty easily, but may need to be updated for non-conflict reasons such as marking other gcc versions `noAbsoluteSoname` (why do these need special handling? do other compilers or runtimes?)

Sure would be great to have this, it's silly for nix to know precisely what libraries to use but every program does a (depending on how you count, n^2) search to rediscover this information.

Anyway I probably don't need to argue for this to the folks on this PR already anyway haha but just a reminder :D.

**Comment by ambrop72** [comment]
I'm interested in getting this further. What is the current status? From what I understand it needs to be rebased and there is a problem with the initrd build.

Libraries that may need to be overridden (e.g. `libGL` for drivers that don't support libglvnd) can still be linked the old way (normal soname and directory in RUNPATH). I also think that the same should be done for `libc` because having more than one instance of libc loaded supposedly is not supported and does not work.

**Comment by lheckemann** [comment]
@ambrop72 I've started to think that this isn't actually a good approach, and patching the `NEEDED` entries may be better. I'm still unsure though.

**Comment by Mic92** [comment]
@lheckemann do you still plan to finish this?

**Comment by lheckemann** [comment]
I definitely still want this to happen — if nobody else beats me to it, I'll have a look at reworking this to patch DT_NEEDED entries At Some Point™.

**Comment by Profpatsch** [comment]
Would love to see us putting absolute paths into `DT_NEEDED`, that looks
like the cleanest solution to me.

**Comment by lheckemann** [comment]
https://github.com/NixOS/patchelf/pull/357 implements setting absolute NEEDED entries, and seems to be quite promising.

**Pull request closed without merging**